### PR TITLE
Better urls for citations

### PR DESCRIPTION
**Pull request by flooie**
PR for better citations

Issue #1599 -- Citation Look Up Tool: return opinion with contained page number

PR provides a better slug/url string for citations, volumes and reporters.
PR also makes citation volumes and reporter pages dynamically private.

### CODE REVIEW

**Review by mlissner — CHANGES_REQUESTED** [review]
This hits the right points. Two things:

1. The slugify code is kind of here and there. Let's consolidate it so it's DRY.
2. We should have some more tests. This feature now has a lot of different ways it can be called, so let's make sure we don't regress and that the new features are tested too. There are already a couple, so I imagine it'll be easy to add more.

**Comment by mlissner** [comment]
Oh, and don't forget we probably need something to go back to volumes from cases, and to reporters from volumes.

File: cl/opinion_page/views.py [hunk 1]
```diff
@@ -642,10 +646,25 @@ def reporter_or_volume_handler(
...
+        (
+            OpinionCluster.objects.filter(
+                citations__reporter=reporter,
+            )
+            .annotate(as_integer=Cast("citations__volume", IntegerField()))
+            .filter(as_integer__range=[volume_int - 1, volume_int + 1])
+            .values_list("as_integer", flat=True)
+            .distinct()
+        )
+    )
```
Comment (mlissner) [inline cl/opinion_page/views.py:664]: Would this work as a query directly on the Citation table instead of doing a join?

Comment (flooie) [reply cl/opinion_page/views.py:664]: 👍  yes it would

File: cl/opinion_page/views.py [hunk 2]
```diff
@@ -789,14 +834,29 @@ def citation_redirector(
...
+                cd["page"] = page
+            return HttpResponseRedirect(
+                reverse("citation_redirector", kwargs=cd)
+            )
+
     # We have a reporter (show volumes in it), a volume (show cases in
     # it), or a citation (show matching citation(s))
+    slug_edition = {slugify(item): item for item in EDITIONS.keys()}
+    proper_reporter = slug_edition[SafeText(reporter)]
+
```
Comment (mlissner) [inline cl/opinion_page/views.py:853]:
```suggestion
    # Look up the slugified reporter to get its proper version (so-2d -> So. 2d)
    slug_edition = {slugify(item): item for item in EDITIONS.keys()}
    proper_reporter = slug_edition[SafeText(reporter)]

    # We have a reporter (show volumes in it), a volume (show cases in
    # it), or a citation (show matching citation(s))
```

**Pull request merged**